flips/20220715-cadence-purity-analysis.md

```diff
@@ -0,0 +1,292 @@
+# Function Purity 
+
+| Status        | (Proposed)       |
+:-------------- |:---------------------------------------------------- |
+| **FLIP #**    | [1056](https://github.com/onflow/flow/pull/1056) |
+| **Author(s)** | Daniel Sainati (daniel.sainati@dapperlabs.com)       |
+| **Sponsor**   | Daniel Sainati (daniel.sainati@dapperlabs.com)       |
+| **Updated**   | 2022-07-15                                           |
+
+## Objective
+
+This FLIP proposes to add a new syntax and accompanying semantic analysis to determine or 
+enforce that a given function or method is "pure", that is, lacking in side effects. This
+would allow additional potential checker rules to enforce immutability in certain contexts, like
+in function conditions or on public fields, that are not possible without a mutability analysis of 
+this sort. 
+
+## Motivation and User Benefit
+
+There are a number of existing issues and proposed features that require or would benefit from 
+the ability to know whether an operation is "pure", or whether it has the potential to modify state. 
+
+The [FLIP to prevent mutation of public fields](https://github.com/onflow/flow/pull/703) attempted 
+to prevent fields in structs and resources, but was only able to prevent mutation of array and dictionary
+fields; nested composites could not be enforced because we did not have a mechanism for knowing what 
+composite methods were "mutating". However, if we could know whether a function or method is `pure`, we
+could restrict the calleable methods on such fields to be only `pure` ones. 
+
+Similarly, a [FLIP](https://github.com/onflow/flow/pull/1043) and an [issue](https://github.com/onflow/cadence/issues/1304)
+that suggest improvements to references to make them less powerful and prone to safety-footguns suggest to limit the 
+operations that can be performed on references, or to allow the creation of read-only references. However, these
+restrictions would only be enforceable at runtime, since we do not have a way statically to decide whether an operation is
+read-only. However, if we could know whether a function or method is `pure`, we could require that such references only 
+perform `pure` operations or are passed to `pure` functions. 
```

Comment:
Would be nice to have some examples of how this FLIP would solve the other issues/FLIPs mentioned in this section